h2o-core/src/main/java/hex/ParallelModelBuilder.java

```diff
@@ -18,13 +19,29 @@
 
   public static abstract class ParallelModelBuilderCallback<D extends ParallelModelBuilderCallback> extends Iced<D> {
 
-    public abstract void onBuildSucces(final Model model, final ParallelModelBuilder parallelModelBuilder);
+    public abstract void onBuildSuccess(final Model model, final ParallelModelBuilder parallelModelBuilder);
+
+    public void postBuildSuccess(final Model model, final ParallelModelBuilder parallelModelBuilder) {};
 
     public abstract void onBuildFailure(final ModelBuildFailure modelBuildFailure, final ParallelModelBuilder parallelModelBuilder);
+
+    public void postBuildFailure(final ModelBuildFailure modelBuildFailure, final ParallelModelBuilder parallelModelBuilder) {};
```

Comment:
do you really need both `postBuildSuccess` and `postBuildFailure`?
Given how they're used in `onModelSuccess` and `onModelFailure`, it looks like they actually have the same goal.
Especially, I don't understand why the callback would want to process the `failure` multiple times, once with `onBuildFailure` and another time with `postBuildFailure`.

I would suggest to merge these 2 `post___` methods to a single `afterCompletion`:
```
public void afterCompletion(final Optional<Model> model, final ParallelModelBuilder parallelModelBuilder);
```

Reply:
(also, I would avoid `post` prefix, at first I thought it was posting something)

Reply:
@seb-h2o this is a quite common naming for a lifecycle kind of things. Let's say in Akka https://alvinalexander.com/scala/understand-methods-akka-actors-scala-lifecycle.

Reply:
@seb-h2o 
> do you really need both postBuildSuccess and postBuildFailure? Given how they're used in onModelSuccess and onModelFailure, it looks like they actually have the same goal.

Not exact the same. Original implementation was dong two things in `onModelSuccess` handler: processing success and making attempt to ask for next model parameters. 

So what I wanted to achieve is to be able to update counters in between of these two actions because I need them to be updated before making decision about whether we want to return next model parameters ( max_models early stopping relies on these counters). There are other ways of how to achieve same result.... but it would also require changes in signatures.

Reply:
> what I wanted to achieve is to be able to update counters in between of these two actions
I understand why you need both `onBuildSuccess` AND `postBuildSuccess`.
But I think you could merge both `postBuildSuccess` AND `postBuildFailure` into a more generic `afterCompletion`.

Reply:
@seb-h2o ah, sorry, my bad. Misunderstood your message.  

`postBuildSuccess` and `postBuildFailure` have different signatures: one has Model parameter and another one has ModelBuilderFailure parameter. Don't think it makes sense to mix these cases together. Having two separate `post...` methods imo also good for a symmetry (otherwise for consistency we could also want generic `onCompletion` method).

Reply:
not convinced that symmetry is good here: as mentioned above, I think that errors (here as wrapped in `ModelBuildFailure`) should be handled only once, i.e. only in `onBuildFailure`, not in `postBuildFailure`, that's why I was suggesting a common `post___` method.
Anyway, don't want to block you more, although this are still potentially public hooks.